appserver/web/web-core/src/main/java/org/apache/catalina/core/StandardContextValve.java

```diff
@@ -325,7 +325,8 @@ private Wrapper preInvoke(Request request, Response response) {
                     || (requestPath.toUpperCase().startsWith("/META-INF/", 0))
                     || (requestPath.equalsIgnoreCase("/META-INF"))
                     || (requestPath.toUpperCase().startsWith("/WEB-INF/", 0))
-                    || (requestPath.equalsIgnoreCase("/WEB-INF"))) {
+                    || (requestPath.equalsIgnoreCase("/WEB-INF"))
```

Comment:
Shouldn't the method normalize above this line have converted to the path to /WEB-INF? What is the output of normalize?

Reply:
I agree. A better solution would be to change the `normalize` method. If the `CoyoteAdapter` class removes the leading `/`, then the `normalize` method should add it at the beginning. The comment on the normalize method also says it normalizes both . and .., but the code only addresses .. and not the . case. The . case seems to be already handled in the `CoyoteAdapter` before this method is called but I believe that the . case should also be addressed here, as the comment says, just in case.

Reply:
I agree that this doesn't seem complete.
If nothing else, this PR as-is doesn't prevent a user going in to the META-INF directory either since a similar curl to `curl -i --path-as-is http://localhost:8080/.//META-INF/MANIFEST.MF` gets normalised to `META-INF/MANIFEST.MF`, which doesn't match any of those _if_ statements (similar case to what you were trying to catch with your extra line here for WEB-INF)

Reply:
> Shouldn't the method normalize above this line have converted to the path to /WEB-INF? What is the output of normalize?

after reviewing in debug mode the CoyoteAdapter is removing the leading / . That is why when this code executes the conditions don't evaluate the final string. For now the approach to improve the normalize method to complete and normalize the URL makes sense. I'm going the changes to improve the normalize method and complete the url with the starting /

Reply:
> I agree. A better solution would be to change the `normalize` method. If the `CoyoteAdapter` class removes the leading `/`, then the `normalize` method should add it at the beginning. The comment on the normalize method also says it normalizes both . and .., but the code only addresses .. and not the . case. The . case seems to be already handled in the `CoyoteAdapter` before this method is called but I believe that the . case should also be addressed here, as the comment says, just in case.

After reviewing the evaluation of the “/../” and “/./” it seems like the process from the CoyoteAdapter is doing most of the work to remove and let the path normalized. Now I added the case to consider the evaluation of the "/./" when I have a path with the following structure:  /app/./some/./something/./my.jsp
then after evaluating the result is:
/app/some/something/my.jsp

![image](https://user-images.githubusercontent.com/279375/131767506-8d3171d4-427b-48f2-a508-be0e332d8e31.png)

let me know if you have questions